### ISSUE TITLE
Host Static Merchant Center App in S3/Google Storage

### ISSUE THREAD

**Issue by Freyert** [commercetools/merchant-center-application-kit#826]
WRT #214 

**Where would you like to deploy your custom application?**
- AWS S3
- Google Storage

**Additional context**
Most custom applications I've seen only need to load a static Single Page Application source code and don't need any server side routing. Only client side routing done within the Single Page Application is required.

While Google Storage offers the ability to "host a static site" there are many limitations that conflict with the current Custom Application bundling process. It's also not a cross platform feature that other clouds implement.

Here's what I think would be nice:

1. Merchant Center Proxy loads `index.html` from my storage bucket.
2. `index.html` loads other resources from bucket.
3. When I add paths off of the Custom Application path these should be a [`window.history.pushstate`](https://stackoverflow.com/questions/3338642/updating-address-bar-with-new-url-without-hash-or-reloading-the-page) event that changes the URL without kicking off any network activity.
4. CSP is defined either in the `index.html` or the configuration is collected from a "well known" file in my S3 bucket.

If it works well for AWS S3 then it will work across a plethora of Clouds as most offer a service that imitates the S3 API.

**Comment by emmenko** [comment]
> Merchant Center Proxy loads `index.html` from my storage bucket.

I haven't thought about that. I can try referencing the storage bucket where the `index.html` is hosted, as the custom app URL, and see what happens. But the proxy should only proxy requests, there should not be any "serving" of files.

> When I add paths off of the Custom Application path these should be a window.history.pushstate event that changes the URL without kicking off any network activity.

I'm not sure I get your point. Can you elaborate this a bit more?

> CSP is defined either in the index.html or the configuration is collected from a "well known" file in my S3 bucket.

There is a proposal to change a bit the way of using the CSP: #293. There are some implications though that we need to fully figure out.
An alternative for now would be to include the CSP into the `index.html`, which is not the best option but it might work for now. However, you still loose some of the security headers (that's what the proposal aims to solve btw).

---

Just for reference, what I also commented in the other issue:

FYI: a quick google search led to me this article, describing how to deploy a SPA to Google Storage.

https://www.verypossible.com/blog/how-to-deploy-a-single-page-application-to-google-cloud

The _First Deployment_ section is the interesting one, to configure the "website".

> In addition, you will need to configure the website configuration. In order to do this, navigate back to the all bucket listing page. You should see your new bucket in the listing table. On the right, click on the three dot menu and then on "Edit website configuration". In the modal that pops up, be sure to enter index.html for the main page and the not found page. The reasoning for this is we are using React Router and want it to handle the not found page.

Also important note mentioned in the article:

> One of the technical limitations of hosting your SPA on a GCP Bucket is your SPA is only accessible via HTTP. In order to protect your users you will need to setup a load balancer so you can serve your users via HTTPS (which will be covered in another blog post).

**Comment by emmenko** [comment]
But anyway, this is definitely a topic that I want to look more into, also for us internally, as a way to simplify deploying and hosting the apps.

I'll get back to you as soon as I have a more concrete solution.

**Comment by Freyert** [comment]
cc: @stephsprinkle

**Comment by emmenko** [comment]
Short update

TL;DR; using a direct link to the bucket won't work because the forwarded requests will not match any of the bucket files. This is a common problem with SPAs, as all requests should be forwarded to the `index.html`.

The problem is that the proxy router forwards the request to the custom app target URL, plus the actual routes of the MC. So if the custom app URL is e.g. `https://my-app.com` and I access the project `my-project/my-custom-app`, the forwarded request would be `https://my-app.com/my-project/my-custom-app`.
Therefore, if the custom app URL is the bucket, the forwarded request to the bucket would be `https://my-bucket/my-project/my-custom-app`, which clearly does not work because there is no file `my-project/my-custom-app`.
Hence the need of e.g. a LB that forwards the requests to the `index.html`.

**Comment by emmenko** [comment]
FYI: @stephsprinkle is working on the documentation related to this topic: #932 

@Freyert would that be enough for you or do you need something else/more?

**Comment by Freyert** [comment]
That works for me! If she's got a process that works for AWS/GCP without using the `mc-http-server` that's perfect.

**Comment by timonag** [comment]
Since pure serverless solution isn't available then could we have **mc-http-server** to work with precompiled _index.html_ rather then building one on each startup?
E.g. GAE does not have writable file system behind thus **mc-http-server** will fail. Also google's autoscaler starts new instances that will cause html compiling.

**Comment by emmenko** [comment]
@timonag sorry for the late answer. I understand your point, can you elaborate though on how you image this to work? For example, should the `index.html` be hosted somewhere else and served by the server? Or do you mount a volume with the `index.html`, then the server reads from the filesystem?

**Comment by timonag** [comment]
It would be ideal to keep `index.html` along with other assets in a google Storage bucket and use bucket's url as an application url when configuring custom application. At least the CSP header is supported by *meta* html tag so it's feasible to compile already secured `index.html`.

For now I've ended with custom Koa http server hosted at GAE that serves `index.html` with headers from a file. Both are compiled during CD pipeline and deployed with `gcloud` and it still seems to me as an unnecessary entities multiplication to say nothing about containers.

**Comment by emmenko** [comment]
> It would be ideal to keep index.html along with other assets in a google Storage bucket and use bucket's url as an application url when configuring custom application.

Well using a direct link to the bucket file won't work as I explained that [here](https://github.com/commercetools/merchant-center-application-kit/issues/826#issuecomment-510123536).
For Single Page Applications you always need some sort of routing in front, to serve all requests from the `index.html`.

If you want to pre-compile the `index.html`, there is a `mc-scripts compile-html` command. Then you can have your little nodejs server that serves the `index.html` from your bucket.

Is that what you are looking for?

**Comment by timonag** [comment]
Not really. Anyway thanks for the response. I'm going to try the DNS way.

**Comment by emmenko** [comment]
Ok, feel free to open a new issue if you have further questions
